src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleBinaryOperator.cs

```diff
@@ -0,0 +1,342 @@
+            var temps = ArrayBuilder<LocalSymbol>.GetInstance();
+
+            BoundExpression newLeft = ReplaceTerminalElementsWithTemps(node.Left, node.Operators, initEffects, temps);
+            BoundExpression newRight = ReplaceTerminalElementsWithTemps(node.Right, node.Operators, initEffects, temps);
+
+            var returnValue = RewriteTupleNestedOperators(node.Operators, newLeft, newRight, boolType, temps, node.OperatorKind);
+            BoundExpression result = MakeSequenceOrResultValue(temps.ToImmutableAndFree(), initEffects.ToImmutableAndFree(), returnValue);
+            return result;
+        }
+
+        private BoundExpression MakeSequenceOrResultValue(ImmutableArray<LocalSymbol> locals, ImmutableArray<BoundExpression> effects, BoundExpression returnValue)
+        {
+            if (locals.IsEmpty && effects.IsEmpty)
+            {
+                return returnValue;
+            }
+
+            return _factory.Sequence(locals, effects, returnValue);
+        }
+
+        /// <summary>
+        /// Walk down tuple literals and replace all the side-effecting elements that need saving with temps.
+        /// Expressions that are not tuple literals need saving, and tuple literals that are involved in a simple comparison rather than a tuple comparison.
+        /// </summary>
+        private BoundExpression ReplaceTerminalElementsWithTemps(BoundExpression expr, TupleBinaryOperatorInfo operators, ArrayBuilder<BoundExpression> initEffects, ArrayBuilder<LocalSymbol> temps)
+        {
+            if (!operators.IsSingle())
+            {
+                // Example:
+                // in `(expr1, expr2) == (..., ...)` we need to save `expr1` and `expr2`
+                if (expr.Kind == BoundKind.TupleLiteral)
+                {
+                    var tuple = (BoundTupleLiteral)expr;
+                    var multiple = (TupleBinaryOperatorInfo.Multiple)operators;
+                    var builder = ArrayBuilder<BoundExpression>.GetInstance(tuple.Arguments.Length);
+                    for (int i = 0; i < tuple.Arguments.Length; i++)
+                    {
+                        var argument = tuple.Arguments[i];
+                        var newArgument = ReplaceTerminalElementsWithTemps(argument, multiple.Operators[i], initEffects, temps);
+                        builder.Add(newArgument);
+                    }
+                    return new BoundTupleLiteral(tuple.Syntax, tuple.ArgumentNamesOpt, tuple.InferredNamesOpt, builder.ToImmutableAndFree(), tuple.Type, tuple.HasErrors);
+                }
+            }
+
+            // Examples:
+            // in `expr == (..., ...)` we need to save `expr` because it's not a tuple literal
+            // in `(..., expr) == (..., (..., ...))` we need to save `expr` because it is used in a simple comparison
+            return EvaluateSideEffectingArgumentToTemp(VisitExpression(expr), initEffects, ref temps);
```

Comment:
>EvaluateSideEffectingArgumentToTemp [](start = 19, length = 35)

In looking at the (existing) implementation of the method `EvaluateSideEffectingArgumentToTemp`, I don't believe it will spill `this` in a value type. I don't know if that was right or wrong before, but that would appear to be incorrect for this use.

Reply:
Good catch. Added test `TestThisStruct` with PROTOTYPE marker

---
In reply to: [172344330](https://github.com/dotnet/roslyn/pull/23356#discussion_r172344330) [](ancestors = 172344330)